upstairs/src/lib.rs

```diff
@@ -1506,56 +1509,152 @@ impl Buffer {
         self.len == 0
     }
 
-    pub async fn as_vec(&self) -> MutexGuard<'_, Vec<u8>> {
-        self.data.lock().await
+    pub fn write(&mut self, offset: usize, data: &[u8]) {
+        assert!(offset + data.len() <= self.data.len());
+
+        self.data[offset..][..data.len()].copy_from_slice(data);
+        self.owned[offset..][..data.len()].fill(true);
+    }
+
+    pub fn write_with_ownership(
+        &mut self,
+        offset: usize,
+        data: &[u8],
+        owned: &[bool],
+    ) {
+        assert!(offset + data.len() <= self.data.len());
+        for i in 0..data.len() {
+            if owned[i] {
+                self.data[offset + i] = data[i];
+                self.owned[offset + i] = true;
+            }
+        }
+    }
+
+    pub fn write_read_response(
+        &mut self,
+        offset: usize,
+        response: &ReadResponse,
+    ) {
+        assert!(offset + response.data.len() <= self.data.len());
+        if !response.block_contexts.is_empty() {
+            self.data[offset..][..response.data.len()]
+                .copy_from_slice(&response.data);
+            self.owned[offset..][..response.data.len()].fill(true);
+        }
+    }
+
+    pub fn read(&self, offset: usize, data: &mut [u8]) {
+        assert!(offset + data.len() <= self.data.len());
+        for i in 0..data.len() {
+            if self.owned[offset + i] {
+                data[i] = self.data[offset + i];
+            }
```

Comment:
This conditional read seems extremely fraught.  (I do a read, but no data is actually copied, but I don't actually know that)  Are we required to keep this around?  It appears to be only used in a test and some test-adjacent (nbd) stuff.

Reply:
There's two reasons we need to keep conditional reads / ownership around: one is that it's essential for the Volume hierarchy layering to work, and two to support `write_unwritten`, which `InMemoryBlockIO` supports for tests (as you noted :)).

In order for Volume hierarchy layering to work, we have to be able to distinguish when the subvolume has been written to in order to know when to layer the subvolume read results on top of what came from the read-only parent. The read-only parent will almost always start out with owned blocks, as the result of a bulk import or snapshot or otherwise, and the subvolumes will start out blank. Ownership is our way of knowing that the zeroes coming from blocks in a subvolume that have never been written to should not override the owned blocks coming from the read-only parent. If there was a write of zeroes to the subvolume, then those _should_ overwrite the blocks from the read-only parent.

It's technically possible to have a read-only parent that was never written to: in this case, `Volume::read` would read no owned blocks from both sides, and the conditional linked to in this comment would never be true.

The second reason is to support `write_unwritten`, but that's not something that occurs in the Upstairs. Receivers of the `write_unwritten` command (i.e. a Downstairs, or `InMemoryBlockIO`) have to determine based on their particular ownership persistence whether or not to write the data: if the block was never written to, then the write goes through. If it was, then the write does not. In the `Downstairs`' case, this is done by querying for existing block context. In `InMemoryBlockIO`'s case, this is done with a `Vec<bool>`.

With authenticated encryption in the mix, we also have the added benefit of only ever returning either all zeroes with no associated authentication, or an authenticated block of data. `owned` in that case means that the read was validated. See `validate_encrypted_read_response` for more on this.